File: modules/cidr/src/main/java/com/falsepattern/zigbrains/cidr/workspace/ZigWorkspaceProvider.java
package com.falsepattern.zigbrains.cidr.workspace;

import com.falsepattern.zigbrains.project.util.ProjectUtil;
import com.intellij.openapi.project.Project;
import com.jetbrains.cidr.project.workspace.CidrWorkspace;
import com.jetbrains.cidr.project.workspace.CidrWorkspaceManager;
import com.jetbrains.cidr.project.workspace.CidrWorkspaceProvider;
import lombok.Cleanup;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.nio.file.Files;

public class ZigWorkspaceProvider implements CidrWorkspaceProvider {
    private @Nullable ZigExternalWorkspace getExistingWorkspace(@NotNull Project project) {
        val workspaces = CidrWorkspaceManager.getInstance(project).getWorkspaces().keySet();
        for (val ws: workspaces) {
            if (ws instanceof ZigExternalWorkspace zew)
                return zew;
        }
        return null;
    }
    @Override
    public @Nullable CidrWorkspace getWorkspace(@NotNull Project project) {
        val existingWorkspace = getExistingWorkspace(project);
        if (existingWorkspace != null) {
            return existingWorkspace;
        }
        val projectDir = ProjectUtil.guessProjectDir(project);
        try {
            @Cleanup val files = Files.walk(projectDir);
            if (files.anyMatch(file -> file.getFileName().toString().endsWith(".zig") || file.getFileName().toString().endsWith(".zig.zon"))) {
                return new ZigExternalWorkspace(project);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public void loadWorkspace(@NotNull Project project) {
        if (getExistingWorkspace(project) != null) {
            return;
        }
        val workspace = getWorkspace(project);
        if (workspace != null) {
            val manager = CidrWorkspaceManager.getInstance(project);
            manager.markInitializing(workspace);
            manager.markInitialized(workspace);
            manager.markLoading(workspace);
            manager.markLoaded(workspace);
        }
    }
}

File: cidr/src/main/kotlin/com/falsepattern/zigbrains/cidr/ZigWorkspaceProvider.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.cidr

import com.falsepattern.zigbrains.shared.sanitizedPathString
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.guessProjectDir
import com.intellij.openapi.vfs.toNioPathOrNull
import com.jetbrains.cidr.project.workspace.CidrWorkspace
import com.jetbrains.cidr.project.workspace.CidrWorkspaceManager
import com.jetbrains.cidr.project.workspace.CidrWorkspaceProvider
import java.io.IOException
import java.nio.file.Files

class ZigWorkspaceProvider: CidrWorkspaceProvider {
    override fun getWorkspace(project: Project): CidrWorkspace? {
        getExistingWorkspace(project)?.let { return it }

        val projectDir = project.guessProjectDir()?.toNioPathOrNull() ?: return null
        try {
            Files.walk(projectDir).use { files ->
                if (files.anyMatch { it.fileName.sanitizedPathString?.let { it2 -> it2.endsWith(".zig") || it2.endsWith(".zig.zon") } ?: false }) {
                    return ZigWorkspace(project)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return null
    }

    override fun loadWorkspace(project: Project) {
        if (getExistingWorkspace(project) != null)
            return
        val workspace = getWorkspace(project)
        if (workspace != null) {
            val manager = CidrWorkspaceManager.getInstance(project)
            manager.markInitializing(workspace)
            manager.markInitialized(workspace)
            manager.markLoading(workspace)
            manager.markLoaded(workspace)
        }
    }
}

private fun getExistingWorkspace(project: Project): ZigWorkspace? {
    val workspaces = CidrWorkspaceManager.getInstance(project).workspaces.keys
    for (ws in workspaces)
        if (ws is ZigWorkspace)
            return ws

    return null
}